finetune/scheduler.py

```diff
@@ -23,89 +60,101 @@ def __init__(self, max_models=None, config=None, reserved=750000000):
+        cpu_percent = psutil.virtual_memory().percent
         LOGGER.info(
             (
                 "models loaded: {num_models}, in_use: {in_use}, max_above_resting: {mar},"
-                " max_model_size: {mms}, gpu_memory_limit: {mem_limit}"
+                " max_model_size: {mms}, gpu_memory_limit: {mem_limit}, cpu percent used: {cpu_percent}"
             ).format(
                 num_models=len(self.loaded_models),
                 in_use=bytes_to_meg(in_use),
                 mar=bytes_to_meg(self.max_above_resting),
                 mms=bytes_to_meg(self.max_model_size),
-                mem_limit=bytes_to_meg(self.gpu_memory_limit)
+                mem_limit=bytes_to_meg(self.gpu_memory_limit),
+                cpu_percent=cpu_percent,
             )
         )
-        return (in_use + self.max_above_resting + self.max_model_size + self.reserved) < self.gpu_memory_limit
+        if cpu_percent > self.ram_max_frac * 100:
+            return False
+        return (
+            in_use + self.max_above_resting + self.max_model_size + self.reserved
+        ) < self.gpu_memory_limit
+
+    def _close_oldest_model(self):
+        name = self.loaded_models.pop(0)
+        self.model_cache[name].close()
+        del self.model_cache[name]
+        gc.collect()
 
     def _rotate_in_model(self, model):
         if model not in self.loaded_models:
             if (
-                    self.max_models is not None and len(self.loaded_models) + 1 > self.max_models or
-                    not self._memory_for_one_more()
+                self.max_models is not None
+                and len(self.loaded_models) + 1 > self.max_models
+                or not self._memory_for_one_more()
             ):
-                name = self.loaded_models.pop(0)
-                self.model_cache[name].close()
-                del self.model_cache[name]
-                gc.collect()
+                self._close_oldest_model()
             out_model = BaseModel.load(model, **self.config)
             self.model_cache[model] = out_model
         else:
             out_model = self.model_cache[model]
-            self.loaded_models.remove(model) # put it back at the end of the queue
-            
+            self.loaded_models.remove(model)  # put it back at the end of the queue
+
         self.loaded_models.append(model)
         out_model._cached_predict = True
-        
+
         return out_model
 
     def _update_memory_limit(self, model):
         if hasattr(model.saver, "variables"):
             del model.saver.variables
             del model.saver.fallback_
-            
+
         if self.session is None:
-            self.session = tf.Session() # delay this so that any options get applied from finetune.
-            self.gpu_memory_limit = self.session.run(tf.contrib.memory_stats.BytesLimit())
+            self.session = (
+                tf.Session()
+            )  # delay this so that any options get applied from finetune.
```

Comment:
Kind of confused why this is in parens

Reply:
Black. I will manually put the comment on the next line.